### PR TITLE
Fix MoE throughput regression: restore expert sharding + stack remat …

### LINKED ISSUE #4005: MoE throughput regression: `activation_batch_moe` missing `'expert'` axis + `minimal_flash` not checkpointing MoE intermediates
#### Problem

I found two independent bugs on `main` that cause a **2.6x throughput regression** for MoE
training on any config using `capacity_factor > 0` (einsum/dense_matmul path) with
expert-parallel sharding as the primary axis (e.g., single-node GPU with
`ici_expert_parallelism=-1`).

**Bug 1 — Missing expert axis (2.5x regression, PRs #3473 / #3606):**
`activation_batch_moe` in `base.yml` maps to `['data', 'fsdp', 'fsdp_transpose']` without
`'expert'`. On configs where `expert` is the only active physical mesh axis (all others
size 1), MoE activations become effectively unsharded. This was likely invisible in upstream
testing because TPU v5p-8 uses `fsdp` as the primary axis, and GPU tests used
`ici_expert_parallelism=1`.

**Bug 2 — Remat checkpoint mismatch (7% regression, PRs #3414 / #3505):**
Checkpoint annotations in `moe.py` were renamed from `mlpwi_*` to `moe_mlpwi_*`, but
`minimal_policy()` in `nnx_decoders.py` was not updated to include the new names. The
`minimal_flash` remat policy silently stops checkpointing MoE intermediates, causing
unnecessary recomputation. PR #3505 partially addressed this but only fixed `types.py`,
not `nnx_decoders.py`.

#### Benchmarks

All results on 1x MI325X node (8 GPUs), DeepSeek V2 Lite 16B, `per_device_batch_size=8`,
`ici_expert_parallelism=-1`, `capacity_factor=1.25`, `remat_policy=minimal_flash`:

| Configuration | TFLOP/s | vs Baseline | Description |
|---|---|---|---|
| [`rocm/maxtext` `release/v26.4`](https://github.com/rocm/maxtext/tree/release/v26.4) (baseline) | 318 | — | Before PRs #3473/#3414; most recent commit includes MoE fix |
| Current `main` (broken) | 120 | **2.6x slower** | Both bugs present |

Note: `release/v26.4` refers to the branch in [`rocm/maxtext`](https://github.com/rocm/maxtext),
which is AMD's fork of `AI-Hypercomputer/maxtext`. The most recent commit on that branch
already contains the MoE sharding fix, making it a known-good baseline.

#### Proposed fix options

I tested five approaches. All fix the primary sharding regression; they differ in how they
handle the remat checkpoint names.

| Option | TFLOP/s | vs Baseline | Sharding fix | Remat fix | Files changed |
|---|---|---|---|---|---|
| A | 324 | 1.02x | Revert `_moe` axis names in `moe.py` | Revert checkpoint names to `mlpwi_*` | `moe.py` |
| B | 361 | 1.13x | Add `'expert'` to `activation_batch_moe` in `base.yml` | Revert checkpoint names to `mlpwi_*` | `base.yml`, `moe.py` |
| B' | 336 | 1.06x | Add `'expert'` to `activation_batch_moe` in `base.yml` | Add `moe_mlpwi_*` to `minimal_policy()` | `base.yml`, `nnx_decoders.py` |
| **B''** | **360** | **1.13x** | **Add `'expert'` to `activation_batch_moe` in `base.yml`** | **Stack both names: inner `mlpwi_*` + outer `moe_mlpwi_*`** | **`base.yml`, `moe.py`** |
| C | 324 | 1.02x | Surgical: only restore `activation_batch` on masks/gate_logits | Revert checkpoint names to `mlpwi_*` | `moe.py` |

**Option A** (324 TFLOP/s): Revert all `_moe` axis names and checkpoint names in `moe.py`.
Full recovery. Discards the axis separation refactor from PRs #3473/#3606.

**Option B** (361 TFLOP/s): Add `'expert'` to `activation_batch_moe` in `base.yml` + revert
checkpoint names in `moe.py`. Best raw performance. Preserves code structure but MoE and
dense MLP share remat names, losing independent remat control from PR #3414.

**Option B'** (336 TFLOP/s): Add `'expert'` to `base.yml` + add `moe_mlpwi_*` names to
`minimal_policy()` in `nnx_decoders.py`. Preserves independent MoE remat control. 7% slower
than B because distinct checkpoint names on shared vs routed experts within MoE layers
prevent XLA from merging their remat schedules.

**Option B'' (recommended)** (360 TFLOP/s): Add `'expert'` to `base.yml` + stack both
checkpoint names in `moe.py` (inner `mlpwi_*` matched by `minimal_flash`, outer
`moe_mlpwi_*` for custom policies). Best performance AND independent MoE remat control.
No changes to `nnx_decoders.py`.

**Option C** (324 TFLOP/s): Only restore `activation_batch` on mask and gate_logit tensors
(pre-dispatch), keeping `activation_batch_moe` on dispatch/mlp axes (post-dispatch). Most
semantically precise but more surgical edits, and 10% slower than B/B''.

#### Notes for the shard_map path

The `shard_map` path (`sparse_matmul`, used when `capacity_factor <= 0`) handles expert
partitioning via `shard_map` itself, so adding `'expert'` to `activation_batch_moe` should
be harmless there — JAX ignores redundant sharding constraints inside `shard_map`. However,
this should be verified on TPU configs that use the shard_map path (e.g., DeepSeek V3 with
`capacity_factor=-1`).

cc @NuojCheng (PRs #3473, #3606) @abhinavgoel95 (PRs #3414, #3505)

### Logs/Output

_No response_

### Environment Information

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by yeandy**
# Description

Fixes https://github.com/AI-Hypercomputer/maxtext/issues/4005

**Changes (Option B''):**

`src/maxtext/configs/base.yml` — 1 line:
```yaml
# Before
['activation_batch_moe', ['data', 'fsdp', 'fsdp_transpose']],
# After
['activation_batch_moe', ['data', 'fsdp', 'fsdp_transpose', 'expert']],
```

`src/maxtext/layers/moe.py` — 9 sites, pattern:
```python
# Before
layer_w0 = adc.checkpoint_name(layer_w0, "moe_mlpwi_0")
# After
layer_w0 = adc.checkpoint_name(adc.checkpoint_name(layer_w0, "mlpwi_0"), "moe_mlpwi_0")
```

# Notes: 
- The `base.yml` change restores expert-axis sharding for MoE activations in the
  einsum/dense_matmul path, fixing the 2.5x regression.
- The stacked `checkpoint_name` pattern ensures `minimal_flash` (which matches `mlpwi_*`)
  correctly checkpoints MoE intermediates via the inner name, while preserving the `moe_*`
  outer name for custom remat policies (PR #3414's original intent).
- No changes to `nnx_decoders.py` or `types.py` required.
- If the maintainers prefer a different option (e.g., Option A to fully revert, or Option C
  for semantic precision), those are also tested and ready.

# Tests

- Benchmarked on MI325X (8 GPU): 120 → 360 TFLOP/s (3x recovery), matching or exceeding
  the `release/v26.4` baseline of 318 TFLOP/s.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by mfrancepillois** [comment]
PR LGTM.  I've also noticed this performance regression on my local setup.

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/base.yml [hunk 1]
```diff
@@ -518,7 +518,7 @@ logical_axis_rules: [
                       # Mixture of Experts (MoE)
                       # ==========================================
                       # MoE Activations
-                      ['activation_batch_moe', ['data', 'fsdp', 'fsdp_transpose']],
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml:521]: Core MoE activations should get sharded on the expert dimension by the expert physical axes, not the batch dimension.

Likely the fix should be instead to modify some constraint with activation_batch instead of activation_batch_moe if it is not a core moe activation (which is kind of subtle, it means the EP comms have not occurred yet)

Comment (gobbleturk) [reply src/maxtext/configs/base.yml:521]: @NuojCheng will work on longterm fix possibly merging this I guess

Comment (NuojCheng) [reply src/maxtext/configs/base.yml:521]: Yes I will work on followup PRs
